apps/common/lib/image.ex

```diff
@@ -0,0 +1,18 @@
+defmodule Common.Image do
+  @moduledoc """
+  Represents a image.
+
+  only `:data` OR `:url` should be set.
+  """
+  alias Common.Image
+
+  defstruct [:data, :url]
+  @typedoc "the typespec of an image"
+  @type t :: %Image{}
+
+  defimpl Common.Validation, for: Image do
```

Comment:
replacing this with a regular function that pattern matches against the input would be way simpler imo:
```elixir
def valid?(%Image{}), do: true
def valid?(_), do: false
```

You could also add guards to validate types (though you'd need to have multiple function heads for optional types since they might not be there)

This'd also solve the recursion question you had, since you can just enforce the struct type in the pattern match

Reply:
The advantage of a protocol is that every module has it's own validation logic in it's .ex file.

Reply:
a struct module can also have functions tho? so you could validate by calling `image |> Image.valid?()` or `mod |> Mod.valid?()`.
instead of a protocol to enforce the existence of validation functions, use a behaviour.
the main advantage of a protocol is that it doesn't care about the data type, but when working with a database struct, you generally already know what data you're dealing with, so thats negated.